x509/src/certificate.rs

```diff
@@ -0,0 +1,243 @@
+                ::der::asn1::ContextSpecific::decode_explicit(decoder, ::der::TagNumber::N3)?
+                    .map(|cs| cs.value);
+            Ok(Self {
+                version,
+                serial_number,
+                signature,
+                issuer,
+                validity,
+                subject,
+                subject_public_key_info,
+                issuer_unique_id,
+                subject_unique_id,
+                extensions,
+            })
+        })
+    }
+}
+const VERSION_TAG: TagNumber = TagNumber::new(0);
+const EXTENSIONS_TAG: TagNumber = TagNumber::new(3);
+impl<'a> ::der::Sequence<'a> for TBSCertificate<'a> {
+    fn fields<F, T>(&self, f: F) -> ::der::Result<T>
+    where
+        F: FnOnce(&[&dyn der::Encodable]) -> ::der::Result<T>,
+    {
+        #[allow(unused_imports)]
+        use core::convert::TryFrom;
+        f(&[
+            &ContextSpecific {
+                tag_number: VERSION_TAG,
+                tag_mode: TagMode::Explicit,
+                value: self.version,
+            },
+            &self.serial_number,
+            &self.signature,
+            &self.issuer,
+            &self.validity,
+            &self.subject,
+            &self.subject_public_key_info,
+            &self.issuer_unique_id,
+            &self.subject_unique_id,
+            &self.extensions.as_ref().map(|exts| ContextSpecific {
+                tag_number: EXTENSIONS_TAG,
+                tag_mode: TagMode::Explicit,
+                value: exts.clone(),
+            }),
+        ])
+    }
+}
+
+impl<'a> ::core::fmt::Debug for TBSCertificate<'a> {
+    fn fmt(&self, f: &mut ::core::fmt::Formatter<'_>) -> ::core::fmt::Result {
+        f.write_fmt(format_args!("\n\tVersion: {:02X?}\n", self.version))?;
+        f.write_fmt(format_args!("\tSerial Number: 0x"))?;
+        for b in self.serial_number.as_bytes() {
+            f.write_fmt(format_args!("{:02X?}", b))?;
+        }
+        f.write_fmt(format_args!("\n"))?;
+        f.write_fmt(format_args!("\tSignature: {:?}\n", self.signature))?;
+        f.write_fmt(format_args!("\tIssuer: {:?}\n", self.issuer))?;
+        f.write_fmt(format_args!("\tValidity: {:?}\n", self.validity))?;
+        f.write_fmt(format_args!("\tSubject: {:?}\n", self.subject))?;
+        f.write_fmt(format_args!(
+            "\tSubject Public Key Info: {:?}\n",
+            self.subject_public_key_info
+        ))?;
+        f.write_fmt(format_args!(
+            "\tIssuer Unique ID: {:?}\n",
+            self.issuer_unique_id
+        ))?;
+        f.write_fmt(format_args!(
+            "\tSubject Unique ID: {:?}\n",
+            self.subject_unique_id
+        ))?;
+        if let Some(exts) = self.extensions.as_ref() {
+            for (i, e) in exts.iter().enumerate() {
+                f.write_fmt(format_args!("\tExtension #{}: {:?}\n", i, e))?;
+            }
+        } else {
+            f.write_fmt(format_args!("\tExtensions: None\n"))?;
+        }
+        Ok(())
```

Comment:
Not something that needs to be addressed in this PR per se, but I think it would make sense to use the [`DebugStruct`](https://doc.rust-lang.org/stable/std/fmt/struct.DebugStruct.html) builder type to build this output, even if some of the fields have customized outputs.

Additionally, it would probably make sense to push the customized output formatting to the `Debug` impls of the individual field types wherever possible.

Reply:
OK. Will address in PR that adds extension decoders and encoders.